File: src/main/java/no/nav/fo/veilarbregistrering/bruker/KontaktinfoService.java
package no.nav.fo.veilarbregistrering.bruker;

import no.nav.fo.veilarbregistrering.bruker.feil.KontaktinfoIngenTilgang;
import no.nav.fo.veilarbregistrering.bruker.feil.KontaktinfoIngenTreff;
import no.nav.fo.veilarbregistrering.bruker.feil.KontaktinfoUkjentFeil;
import no.nav.fo.veilarbregistrering.feil.FeilType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.ws.rs.ForbiddenException;
import javax.ws.rs.NotAuthorizedException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class KontaktinfoService {

    private static final Logger LOG = LoggerFactory.getLogger(KontaktinfoService.class);

    private final PdlOppslagGateway pdlOppslagGateway;
    private final KrrGateway krrGateway;

    public KontaktinfoService(PdlOppslagGateway pdlOppslagGateway, KrrGateway krrGateway) {
        this.pdlOppslagGateway = pdlOppslagGateway;
        this.krrGateway = krrGateway;
    }

    public Kontaktinfo hentKontaktinfo(Bruker bruker) {

        List<FeilType> feiltyper = new ArrayList<>(2);

        Optional<Person> person;
        try {
            person = pdlOppslagGateway.hentPerson(bruker.getAktorId());
        } catch (RuntimeException e) {
            LOG.error("Hent kontaktinfo fra PDL feilet", e);
            person = Optional.empty();
            feiltyper.add(FeilType.UKJENT);
        }

        Optional<Telefonnummer> telefonnummer;
        try {
            telefonnummer = krrGateway.hentKontaktinfo(bruker);

        } catch (NotAuthorizedException | ForbiddenException e) {
            LOG.error("Hent kontaktinfo fra Kontakt og reservasjonsregisteret feilet pga manglende tilgang", e);
            telefonnummer = Optional.empty();
            feiltyper.add(FeilType.INGEN_TILGANG);

        } catch (RuntimeException e) {
            LOG.error("Hent kontaktinfo fra Kontakt og reservasjonsregisteret feilet av ukjent grunn", e);
            telefonnummer = Optional.empty();
            feiltyper.add(FeilType.UKJENT);
        }

        if (fantMinstEttTelefonnummer(person, telefonnummer)) {
            return opprettKontaktinfo(person, telefonnummer);
        }

        if (feiltyper.contains(FeilType.INGEN_TILGANG)) {
            throw new KontaktinfoIngenTilgang();
        }
        if (feiltyper.contains(FeilType.UKJENT)) {
            throw new KontaktinfoUkjentFeil();
        }
        throw new KontaktinfoIngenTreff();
    }

    private boolean fantMinstEttTelefonnummer(Optional<Person> person, Optional<Telefonnummer> telefonnummer) {
        return (person.isPresent() && person.get().getTelefonnummer().isPresent())
                || telefonnummer.isPresent();
    }

    private Kontaktinfo opprettKontaktinfo(Optional<Person> person, Optional<Telefonnummer> telefonnummer) {
        return Kontaktinfo.of(
                person.flatMap(p -> p.getTelefonnummer()
                        .map(Telefonnummer::asLandkodeOgNummer))
                        .orElse(null),
                telefonnummer
                        .orElse(null)
        );
    }
}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/bruker/KontaktinfoService.kt
package no.nav.fo.veilarbregistrering.bruker

import no.nav.fo.veilarbregistrering.bruker.feil.KontaktinfoIngenTilgang
import no.nav.fo.veilarbregistrering.bruker.feil.KontaktinfoIngenTreff
import no.nav.fo.veilarbregistrering.bruker.feil.KontaktinfoUkjentFeil
import no.nav.fo.veilarbregistrering.feil.FeilType
import org.slf4j.LoggerFactory
import javax.ws.rs.ForbiddenException
import javax.ws.rs.NotAuthorizedException

class KontaktinfoService(private val pdlOppslagGateway: PdlOppslagGateway, private val krrGateway: KrrGateway) {
    fun hentKontaktinfo(bruker: Bruker): Kontaktinfo {
        val feiltyper: MutableList<FeilType> = ArrayList(2)
        val person: Person? = try {
            pdlOppslagGateway.hentPerson(bruker.aktorId)
        } catch (e: RuntimeException) {
            LOG.error("Hent kontaktinfo fra PDL feilet", e)
            feiltyper.add(FeilType.UKJENT)
            null
        }
        val telefonnummer: Telefonnummer? = try {
            krrGateway.hentKontaktinfo(bruker)
        } catch (e: NotAuthorizedException) {
            LOG.error("Hent kontaktinfo fra Kontakt og reservasjonsregisteret feilet pga manglende autentisering", e)
            feiltyper.add(FeilType.INGEN_TILGANG)
            null
        } catch (e: ForbiddenException) {
            LOG.error("Hent kontaktinfo fra Kontakt og reservasjonsregisteret feilet pga manglende tilgang", e)
            feiltyper.add(FeilType.INGEN_TILGANG)
            null
        } catch (e: RuntimeException) {
            LOG.error("Hent kontaktinfo fra Kontakt og reservasjonsregisteret feilet av ukjent grunn", e)
            feiltyper.add(FeilType.UKJENT)
            null
        }
        if (fantMinstEttTelefonnummer(person, telefonnummer)) {
            return opprettKontaktinfo(person, telefonnummer)
        }
        if (feiltyper.contains(FeilType.INGEN_TILGANG)) {
            throw KontaktinfoIngenTilgang("Ingen tilgang ved kall til PDL eller KRR")
        }
        if (feiltyper.contains(FeilType.UKJENT)) {
            throw KontaktinfoUkjentFeil("Ukjent feil ved henting av kontaktinfo fra PDL eller KRR")
        }
        throw KontaktinfoIngenTreff("Ingen treff ved oppslag i PDL eller KRR")
    }

    private fun fantMinstEttTelefonnummer(person: Person?, telefonnummer: Telefonnummer?): Boolean {
        return (person?.telefonnummer?.nummer != null)
                || telefonnummer != null
    }

    private fun opprettKontaktinfo(person: Person?, telefonnummer: Telefonnummer?): Kontaktinfo {
        return Kontaktinfo(
            person?.telefonnummer
                ?.let { obj: Telefonnummer -> obj.asLandkodeOgNummer() },
            telefonnummer,
            person?.navn
        )
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(KontaktinfoService::class.java)
    }
}